language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionNodeAnalyzer.java

```diff
@@ -16,74 +16,117 @@
 import io.ballerina.compiler.syntax.tree.NodeVisitor;
+import io.ballerina.compiler.syntax.tree.NonTerminalNode;
+import io.ballerina.compiler.syntax.tree.ObjectTypeDescriptorNode;
+import io.ballerina.compiler.syntax.tree.RecordTypeDescriptorNode;
 import io.ballerina.compiler.syntax.tree.ServiceDeclarationNode;
+import io.ballerina.compiler.syntax.tree.StatementNode;
 import io.ballerina.compiler.syntax.tree.SyntaxKind;
+import io.ballerina.compiler.syntax.tree.SyntaxTree;
 import io.ballerina.compiler.syntax.tree.Token;
 import io.ballerina.compiler.syntax.tree.TypeDefinitionNode;
-import org.ballerinalang.langserver.codeaction.NodeBasedPositionDetailsImpl.PositionDetailsBuilder;
+import io.ballerina.compiler.syntax.tree.VariableDeclarationNode;
+import io.ballerina.tools.text.LinePosition;
+import org.ballerinalang.langserver.common.utils.CommonUtil;
+import org.ballerinalang.langserver.commons.codeaction.CodeActionNodeType;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.Range;
 
 import java.util.Optional;
 
 /**
- * Node analyzer for the code actions.
- * This will calculate the node information and fill the {@link PositionDetailsBuilder} and the information will be
- * reflected to the code actions via the PositionDetails API.
+ * Node analyzer for the code actions. This analyzer will determine
+ * 1. the node for node based code actions
+ * 2. Code action node type
+ * 3. Closest statement node
+ * 4. Documentable node
+ * 5. Enclosing documentable node
  *
- * @since 2201.1.x
+ * @since 2201.1.0
  */
 public class CodeActionNodeAnalyzer extends NodeVisitor {
-    private final PositionDetailsBuilder positionDetailsBuilder;
-    private final int positionOffset;
 
-    public CodeActionNodeAnalyzer(PositionDetailsBuilder positionDetailsBuilder, int positionOffset) {
-        this.positionDetailsBuilder = positionDetailsBuilder;
-        this.positionOffset = positionOffset;
+    private int positionOffset;
+
+    private NonTerminalNode codeActionNode;
+    private CodeActionNodeType codeActionNodeType;
+    private StatementNode statementNode;
+    private NonTerminalNode documentableNode;
+    private NonTerminalNode enclosingDocumentableNode;
+
+    public CodeActionNodeAnalyzer() {
+    }
+
+    /**
+     * Entry point to the analyzer. Provided the cursor position and the syntax tree, this method will calculate the
+     * node details as mentioned in the class level docs.
+     *
+     * @param cursorPosition Cursor position
+     * @param syntaxTree     Syntax tree
+     */
+    public void analyze(Position cursorPosition, SyntaxTree syntaxTree) {
+        LinePosition linePos = LinePosition.from(cursorPosition.getLine(), cursorPosition.getCharacter());
+        this.positionOffset = syntaxTree.textDocument().textPositionFrom(linePos);
+        NonTerminalNode node = CommonUtil.findNode(new Range(cursorPosition, cursorPosition), syntaxTree);
+        if (node.kind() == SyntaxKind.LIST) {
```

Comment:
What is the reason to special case this?

Reply:
Because the `accept()` method is not implemented for `ExternalTreeNodeList` node.